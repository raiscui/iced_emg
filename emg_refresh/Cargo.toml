[package]
name = "emg_refresh"
version = "0.1.0"
authors = ["崔路明 <cuiluming@chaoway.net>"]
edition = "2021"
resolver = "2"
# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
[profile.dev]
split-debuginfo = "unpacked"
[lib]
crate-type = ["cdylib", "rlib"]

[dependencies]
emg_state = { path = "../emg_state" }
emg_core = { path = "../emg_core" }

tracing = "0.1"

[target.'cfg(target_arch = "wasm32")'.dependencies]
# console_log = { version = "0.2", features = ["color"] }


[dev-dependencies]
wasm-bindgen-test = "0.3"
anchors = { path = "../../anchors" }
console_error_panic_hook = "0.1"
cfg-if = "1"
# ctor = "0.1.20"
[target.'cfg(target_arch = "wasm32")'.dev-dependencies]
tracing-wasm = "0.2.0"
