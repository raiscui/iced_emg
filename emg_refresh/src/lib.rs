#![deny(clippy::all)]
#![deny(clippy::pedantic)]
#![warn(clippy::nursery)]
// ────────────────────────────────────────────────────────────────────────────────
// #![feature(specialization)]
#![feature(min_specialization)]
#![feature(auto_traits)]
#![feature(negative_impls)]
// #![feature(trivial_bounds)]
#[cfg(test)]
mod test;
// ────────────────────────────────────────────────────────────────────────────────

mod impl_refresh;
mod refresh_use;
mod refresh_use_for;
mod refreshers;
pub use impl_refresh::RefreshUseNoWarper;
pub use impl_refresh::RefreshWhoNoWarper;
pub use refresh_use::RefreshUse;
pub use refresh_use_for::RefreshForUse;
pub use refreshers::EqRefreshFor;
pub use refreshers::RefreshFor;
pub use refreshers::Refresher;
pub use refreshers::RefresherFor;

#[cfg(test)]
mod tests {
    #[test]
    fn it_works() {
        let x = 2 + 2;
        assert_eq!(x, 4);
    }
}
