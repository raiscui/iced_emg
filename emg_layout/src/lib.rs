#![deny(clippy::all)]
#![deny(clippy::pedantic)]
#![warn(clippy::nursery)]
// #![feature(negative_impls)]
// #![feature(auto_traits)]

use derive_more::Display;
use derive_more::From;
use emg_refresh::{RefreshFor, RefreshUseFor, RefreshUseNoWarper, RefreshWhoNoWarper};
use na::{Affine3, Isometry3, Matrix4, Point3, Rotation3, Similarity3, Translation3, Vector3};
use nalgebra as na;
pub use seed_styles as styles;
use std::rc::Rc;
use styles::{pc, px, CssValueTrait, ExactLength, Percent, Style, UpdateStyle};

type Vec3 = Vector3<f64>;
type Mat4 = Matrix4<f64>;
type Trans3 = Translation3<f64>;
type Rot3 = Rotation3<f64>;
type Transform9 = Affine3<f64>;
type pos = Point3<f64>;

#[derive(Display, Clone, Debug)]
#[display(fmt = "width: {};")]
pub enum GenericSize {
    #[display(fmt = "auto")]
    Auto,
    Length(ExactLength),
    Percentage(Percent),
    #[display(fmt = "initial")]
    Initial,
    #[display(fmt = "inherit")]
    Inherit,
    StringValue(String),
}
impl Default for GenericSize {
    fn default() -> Self {
        Self::Length(px(0))
    }
}

impl From<ExactLength> for GenericSize {
    fn from(length: ExactLength) -> Self {
        Self::Length(length)
    }
}
impl From<Percent> for GenericSize {
    fn from(percentage: Percent) -> Self {
        Self::Percentage(percentage)
    }
}
#[derive(Debug, Clone)]
struct WHSize {
    w: GenericSize,
    h: GenericSize,
}
impl Default for WHSize {
    fn default() -> Self {
        Self {
            w: px(16).into(),
            h: px(16).into(),
        }
    }
}

#[derive(Debug, Clone)]
struct Layout {
    size: WHSize,
    origin: GenericSize,
    align: GenericSize,
}

impl Default for Layout {
    fn default() -> Self {
        Self {
            size: Default::default(),
            origin: pc(0).into(),
            align: pc(0).into(),
        }
    }
}

#[derive(Debug, Clone)]
struct LayoutCompiled {
    w_either: Option<GenericSize>,
    origin_x_either: Option<GenericSize>,
    align_x_either: Option<GenericSize>,
    h_either: Option<GenericSize>,
    origin_y_either: Option<GenericSize>,
    align_y_either: Option<GenericSize>,
}
impl Default for LayoutCompiled {
    fn default() -> Self {
        Self {
            w_either: None,
            origin_x_either: None,
            align_x_either: None,
            h_either: None,
            origin_y_either: None,
            align_y_either: None,
        }
    }
}

struct Transforms {
    loc: Trans3,
    scale: Vec3,
    rotate: Rot3,
}
impl Default for Transforms {
    fn default() -> Self {
        Self {
            loc: Trans3::identity(),
            scale: Vec3::from_element(0.),
            rotate: Rot3::identity(),
        }
    }
}
#[derive(Debug, Clone)]
struct M4Data {
    m4: Transform9,
    m4_def: Transform9,
    layout_m4: Transform9,
    //TODO m4fDef:ED msg -> M4.Mat4
    world_inv: Transform9,
    layout_inv: Transform9,
    m4offset: Transform9,
}
impl Default for M4Data {
    fn default() -> Self {
        Self {
            m4: Transform9::identity(),
            m4_def: Transform9::identity(),
            layout_m4: Transform9::identity(),
            //TODO m4fDef:ED msg -> M4.Mat4
            world_inv: Transform9::identity(),
            layout_inv: Transform9::identity(),
            m4offset: Transform9::identity(),
        }
    }
}

#[derive(Default, Debug, Clone)]
struct EdgeDataOutput {
    loc_styles: Style,
}
#[derive(Default, Debug, Clone)]
pub struct EdgeData {
    id: String,
    layout: Layout,
    compiled: LayoutCompiled,
    matrix: M4Data,
    // transforms_am: Transforms,
    ed_output: EdgeDataOutput,
    // animations:
}
#[derive(From, Clone, Debug)]
pub enum EdgeItem {
    EdgeData(EdgeData),
    String(String),
    Empty,
}

impl Default for EdgeItem {
    fn default() -> Self {
        Self::Empty
    }
}
pub fn e() -> EdgeData {
    EdgeData::default()
}

// TODO lifetime
#[derive(Clone)]
struct Css<T>(T)
where
    T: CssValueTrait + Clone + 'static;

impl<T: std::clone::Clone + seed_styles::CssValueTrait> From<T> for Css<T> {
    fn from(v: T) -> Self {
        Self(v)
    }
}

// impl RefreshWhoNoWarper for EdgeData {}
// TODO lifetime
impl RefreshFor<EdgeData> for Vec<Box<(dyn RefreshFor<EdgeData> + 'static)>> {
    fn refresh_for(&self, who: &mut EdgeData) {
        for i in self {
            let ii = i.as_ref();
            who.refresh_use(ii);
        }
    }
}
impl<Use> RefreshFor<EdgeData> for Css<Use>
where
    Use: CssValueTrait + std::clone::Clone,
{
    fn refresh_for(&self, who: &mut EdgeData) {
        let t = self.0.clone();
        t.update_style(&mut who.ed_output.loc_styles);
    }
}
//TODO lifetime
pub fn css<Use: CssValueTrait + std::clone::Clone + 'static>(v: Use) -> impl RefreshFor<EdgeData> {
    Css(v)
}
// impl<Use> RefreshFor<EdgeData> for Css<Use>
// where
//     Use: CssValueTrait + std::clone::Clone,
// {
//     fn refresh_for(&self, who: &mut EdgeData) {
//         let t = self.0.clone();
//         t.update_style(&mut who.ed_output.loc_styles);
//     }
// }
impl RefreshFor<EdgeData> for Style {
    fn refresh_for(&self, who: &mut EdgeData) {
        who.ed_output.loc_styles = self.clone();
    }
}
impl RefreshFor<EdgeData> for Layout {
    fn refresh_for(&self, who: &mut EdgeData) {
        who.layout = self.clone();
    }
}

#[cfg(test)]
mod tests {

    use seed_styles::CssWidth;

    use super::*;

    #[test]
    fn it_works() {
        use seed_styles::CssWidthTrait;

        println!("{:?}", Transform9::identity());
        let cc = Transform9::identity();
        let p = px(2);
        let cc = CssWidth::from(px(2));
        let mut e = EdgeData::default();
        e.refresh_use(&Css(cc));
        println!("{:?}", &e);
        let mut s = Style::default();
        s = s.style_child("dddddd");
        println!("{:?}", &s);
        println!("{:?}", &s.render());
        // println!("{:?}", cc.inverse());
        // println!("{:?}", &mut s.w(px(11)).render());
        // println!("{:#?}", e());
        // ─────────────────────────────────────────────────────────────────
    }
}
