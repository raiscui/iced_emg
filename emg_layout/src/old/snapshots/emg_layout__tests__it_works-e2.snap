---
source: emg_layout/src/lib.rs
expression: "&e2"

---
EmgEdgeItem {
    id: StateVar(
        StateAnchor(
            EdgeIndex(Some(NodeIndex("1"))->Some(NodeIndex("2"))),
        ),
    ),
    paths: StateAnchor(
        {
            EPath(
                [
                    EdgeIndex(None->Some(NodeIndex("root"))),
                    EdgeIndex(Some(NodeIndex("root"))->Some(NodeIndex("1"))),
                ],
            ): EdgeData(
                EdgeData {
                    calculated: LayoutCalculated {
                        size: StateAnchor(
                            Matrix {
                                data: [
                                    [
                                        12.0,
                                        200.0,
                                    ],
                                ],
                            },
                        ),
                        origin: StateAnchor(
                            Translation {
                                vector: Matrix {
                                    data: [
                                        [
                                            -12.0,
                                            -200.0,
                                            0.0,
                                        ],
                                    ],
                                },
                            },
                        ),
                        align: StateAnchor(
                            Translation {
                                vector: Matrix {
                                    data: [
                                        [
                                            50.0,
                                            20.0,
                                            0.0,
                                        ],
                                    ],
                                },
                            },
                        ),
                        coordinates_trans: StateAnchor(
                            Translation {
                                vector: Matrix {
                                    data: [
                                        [
                                            38.0,
                                            -180.0,
                                            0.0,
                                        ],
                                    ],
                                },
                            },
                        ),
                        matrix: StateAnchor(
                            Mat4(
                                Matrix {
                                    data: [
                                        [
                                            1.0,
                                            0.0,
                                            0.0,
                                            0.0,
                                        ],
                                        [
                                            0.0,
                                            1.0,
                                            0.0,
                                            0.0,
                                        ],
                                        [
                                            0.0,
                                            0.0,
                                            1.0,
                                            0.0,
                                        ],
                                        [
                                            38.0,
                                            -180.0,
                                            0.0,
                                            1.0,
                                        ],
                                    ],
                                },
                            ),
                        ),
                        loc_styles: StateAnchor(
                            Style {
                                media_rules: {},
                                rules: [
                                    Rule {
                                        value: Length(
                                            Simplex(
                                                ExactLengthSimplex {
                                                    unit: Px,
                                                    value: NotNan(
                                                        12.0,
                                                    ),
                                                },
                                            ),
                                        ),
                                    },
                                    Rule {
                                        value: Length(
                                            Simplex(
                                                ExactLengthSimplex {
                                                    unit: Px,
                                                    value: NotNan(
                                                        200.0,
                                                    ),
                                                },
                                            ),
                                        ),
                                    },
                                    Rule {
                                        value: StringValue(
                                            "matrix3d(1,0,0,0,0,1,0,0,0,0,1,0,38,-180,0,1)",
                                        ),
                                    },
                                ],
                                updated_at: [
                                    "emg_layout/src/calc.rs:153:29",
                                    "emg_layout/src/calc.rs:153:33",
                                    "emg_layout/src/calc.rs:153:52",
                                    "emg_layout/src/calc.rs:153:71",
                                ],
                                pseudo: None,
                                media: None,
                                name: "",
                                keyframes: Keyframes {
                                    frames: [],
                                },
                                combinator: None,
                                pre_combinators: [],
                            },
                        ),
                    },
                    styles_string: StateAnchor(
                        "width: 12px;\nheight: 200px;\ntransform: matrix3d(1,0,0,0,0,1,0,0,0,0,1,0,38,-180,0,1);\n",
                    ),
                    opt_p_calculated: Some(
                        LayoutCalculated {
                            size: StateAnchor(
                                Matrix {
                                    data: [
                                        [
                                            100.0,
                                            100.0,
                                        ],
                                    ],
                                },
                            ),
                            origin: StateAnchor(
                                Translation {
                                    vector: Matrix {
                                        data: [
                                            [
                                                0.0,
                                                0.0,
                                                0.0,
                                            ],
                                        ],
                                    },
                                },
                            ),
                            align: StateAnchor(
                                Translation {
                                    vector: Matrix {
                                        data: [
                                            [
                                                0.0,
                                                0.0,
                                                0.0,
                                            ],
                                        ],
                                    },
                                },
                            ),
                            coordinates_trans: StateAnchor(
                                Translation {
                                    vector: Matrix {
                                        data: [
                                            [
                                                0.0,
                                                0.0,
                                                0.0,
                                            ],
                                        ],
                                    },
                                },
                            ),
                            matrix: StateAnchor(
                                Mat4(
                                    Matrix {
                                        data: [
                                            [
                                                1.0,
                                                0.0,
                                                0.0,
                                                0.0,
                                            ],
                                            [
                                                0.0,
                                                1.0,
                                                0.0,
                                                0.0,
                                            ],
                                            [
                                                0.0,
                                                0.0,
                                                1.0,
                                                0.0,
                                            ],
                                            [
                                                0.0,
                                                0.0,
                                                0.0,
                                                1.0,
                                            ],
                                        ],
                                    },
                                ),
                            ),
                            loc_styles: StateAnchor(
                                Style {
                                    media_rules: {},
                                    rules: [
                                        Rule {
                                            value: Length(
                                                Simplex(
                                                    ExactLengthSimplex {
                                                        unit: Px,
                                                        value: NotNan(
                                                            100.0,
                                                        ),
                                                    },
                                                ),
                                            ),
                                        },
                                        Rule {
                                            value: Length(
                                                Simplex(
                                                    ExactLengthSimplex {
                                                        unit: Px,
                                                        value: NotNan(
                                                            100.0,
                                                        ),
                                                    },
                                                ),
                                            ),
                                        },
                                        Rule {
                                            value: StringValue(
                                                "matrix3d(1,0,0,0,0,1,0,0,0,0,1,0,0,0,0,1)",
                                            ),
                                        },
                                    ],
                                    updated_at: [
                                        "emg_layout/src/lib.rs:958:13",
                                        "emg_layout/src/lib.rs:958:17",
                                        "emg_layout/src/lib.rs:958:31",
                                        "emg_layout/src/lib.rs:958:45",
                                    ],
                                    pseudo: None,
                                    media: None,
                                    name: "",
                                    keyframes: Keyframes {
                                        frames: [],
                                    },
                                    combinator: None,
                                    pre_combinators: [],
                                },
                            ),
                        },
                    ),
                },
            ),
        },
    ),
    layout: Layout {
        w: StateVar(
            GenericSizeAnchor(
                StateAnchor(
                    Length(
                        Simplex(
                            ExactLengthSimplex {
                                unit: Px,
                                value: NotNan(
                                    100.0,
                                ),
                            },
                        ),
                    ),
                ),
            ),
        ),
        h: StateVar(
            GenericSizeAnchor(
                StateAnchor(
                    Length(
                        Simplex(
                            ExactLengthSimplex {
                                unit: Px,
                                value: NotNan(
                                    150.0,
                                ),
                            },
                        ),
                    ),
                ),
            ),
        ),
        z: StateVar(
            StateAnchor(
                0,
            ),
        ),
        origin_x: StateVar(
            GenericSizeAnchor(
                StateAnchor(
                    Length(
                        Simplex(
                            ExactLengthSimplex {
                                unit: Pc,
                                value: NotNan(
                                    100.0,
                                ),
                            },
                        ),
                    ),
                ),
            ),
        ),
        origin_y: StateVar(
            GenericSizeAnchor(
                StateAnchor(
                    Length(
                        Simplex(
                            ExactLengthSimplex {
                                unit: Pc,
                                value: NotNan(
                                    100.0,
                                ),
                            },
                        ),
                    ),
                ),
            ),
        ),
        origin_z: StateVar(
            GenericSizeAnchor(
                StateAnchor(
                    Length(
                        Simplex(
                            ExactLengthSimplex {
                                unit: Pc,
                                value: NotNan(
                                    100.0,
                                ),
                            },
                        ),
                    ),
                ),
            ),
        ),
        align_x: StateVar(
            GenericSizeAnchor(
                StateAnchor(
                    Length(
                        Simplex(
                            ExactLengthSimplex {
                                unit: Pc,
                                value: NotNan(
                                    50.0,
                                ),
                            },
                        ),
                    ),
                ),
            ),
        ),
        align_y: StateVar(
            GenericSizeAnchor(
                StateAnchor(
                    Length(
                        Simplex(
                            ExactLengthSimplex {
                                unit: Pc,
                                value: NotNan(
                                    20.0,
                                ),
                            },
                        ),
                    ),
                ),
            ),
        ),
        align_z: StateVar(
            GenericSizeAnchor(
                StateAnchor(
                    Length(
                        Simplex(
                            ExactLengthSimplex {
                                unit: Pc,
                                value: NotNan(
                                    20.0,
                                ),
                            },
                        ),
                    ),
                ),
            ),
        ),
    },
    path_styles: StateVar(
        {},
    ),
    path_layouts: StateVar(
        {},
    ),
    other_styles: StateVar(
        Style {
            media_rules: {},
            rules: [],
            updated_at: [
                "emg_layout/src/lib.rs:731:41",
            ],
            pseudo: None,
            media: None,
            name: "",
            keyframes: Keyframes {
                frames: [],
            },
            combinator: None,
            pre_combinators: [],
        },
    ),
    edge_nodes: StateAnchor(
        {
            EPath(
                [
                    EdgeIndex(None->Some(NodeIndex("root"))),
                    EdgeIndex(Some(NodeIndex("root"))->Some(NodeIndex("1"))),
                    EdgeIndex(Some(NodeIndex("1"))->Some(NodeIndex("2"))),
                ],
            ): EdgeData(
                EdgeData {
                    calculated: LayoutCalculated {
                        size: StateAnchor(
                            Matrix {
                                data: [
                                    [
                                        100.0,
                                        150.0,
                                    ],
                                ],
                            },
                        ),
                        origin: StateAnchor(
                            Translation {
                                vector: Matrix {
                                    data: [
                                        [
                                            -100.0,
                                            -150.0,
                                            0.0,
                                        ],
                                    ],
                                },
                            },
                        ),
                        align: StateAnchor(
                            Translation {
                                vector: Matrix {
                                    data: [
                                        [
                                            6.0,
                                            40.0,
                                            0.0,
                                        ],
                                    ],
                                },
                            },
                        ),
                        coordinates_trans: StateAnchor(
                            Translation {
                                vector: Matrix {
                                    data: [
                                        [
                                            -94.0,
                                            -110.0,
                                            0.0,
                                        ],
                                    ],
                                },
                            },
                        ),
                        matrix: StateAnchor(
                            Mat4(
                                Matrix {
                                    data: [
                                        [
                                            1.0,
                                            0.0,
                                            0.0,
                                            0.0,
                                        ],
                                        [
                                            0.0,
                                            1.0,
                                            0.0,
                                            0.0,
                                        ],
                                        [
                                            0.0,
                                            0.0,
                                            1.0,
                                            0.0,
                                        ],
                                        [
                                            -94.0,
                                            -110.0,
                                            0.0,
                                            1.0,
                                        ],
                                    ],
                                },
                            ),
                        ),
                        loc_styles: StateAnchor(
                            Style {
                                media_rules: {},
                                rules: [
                                    Rule {
                                        value: Length(
                                            Simplex(
                                                ExactLengthSimplex {
                                                    unit: Px,
                                                    value: NotNan(
                                                        100.0,
                                                    ),
                                                },
                                            ),
                                        ),
                                    },
                                    Rule {
                                        value: Length(
                                            Simplex(
                                                ExactLengthSimplex {
                                                    unit: Px,
                                                    value: NotNan(
                                                        150.0,
                                                    ),
                                                },
                                            ),
                                        ),
                                    },
                                    Rule {
                                        value: StringValue(
                                            "matrix3d(1,0,0,0,0,1,0,0,0,0,1,0,-94,-110,0,1)",
                                        ),
                                    },
                                ],
                                updated_at: [
                                    "emg_layout/src/calc.rs:153:29",
                                    "emg_layout/src/calc.rs:153:33",
                                    "emg_layout/src/calc.rs:153:52",
                                    "emg_layout/src/calc.rs:153:71",
                                ],
                                pseudo: None,
                                media: None,
                                name: "",
                                keyframes: Keyframes {
                                    frames: [],
                                },
                                combinator: None,
                                pre_combinators: [],
                            },
                        ),
                    },
                    styles_string: StateAnchor(
                        "width: 100px;\nheight: 150px;\ntransform: matrix3d(1,0,0,0,0,1,0,0,0,0,1,0,-94,-110,0,1);\n",
                    ),
                    opt_p_calculated: Some(
                        LayoutCalculated {
                            size: StateAnchor(
                                Matrix {
                                    data: [
                                        [
                                            12.0,
                                            200.0,
                                        ],
                                    ],
                                },
                            ),
                            origin: StateAnchor(
                                Translation {
                                    vector: Matrix {
                                        data: [
                                            [
                                                -12.0,
                                                -200.0,
                                                0.0,
                                            ],
                                        ],
                                    },
                                },
                            ),
                            align: StateAnchor(
                                Translation {
                                    vector: Matrix {
                                        data: [
                                            [
                                                50.0,
                                                20.0,
                                                0.0,
                                            ],
                                        ],
                                    },
                                },
                            ),
                            coordinates_trans: StateAnchor(
                                Translation {
                                    vector: Matrix {
                                        data: [
                                            [
                                                38.0,
                                                -180.0,
                                                0.0,
                                            ],
                                        ],
                                    },
                                },
                            ),
                            matrix: StateAnchor(
                                Mat4(
                                    Matrix {
                                        data: [
                                            [
                                                1.0,
                                                0.0,
                                                0.0,
                                                0.0,
                                            ],
                                            [
                                                0.0,
                                                1.0,
                                                0.0,
                                                0.0,
                                            ],
                                            [
                                                0.0,
                                                0.0,
                                                1.0,
                                                0.0,
                                            ],
                                            [
                                                38.0,
                                                -180.0,
                                                0.0,
                                                1.0,
                                            ],
                                        ],
                                    },
                                ),
                            ),
                            loc_styles: StateAnchor(
                                Style {
                                    media_rules: {},
                                    rules: [
                                        Rule {
                                            value: Length(
                                                Simplex(
                                                    ExactLengthSimplex {
                                                        unit: Px,
                                                        value: NotNan(
                                                            12.0,
                                                        ),
                                                    },
                                                ),
                                            ),
                                        },
                                        Rule {
                                            value: Length(
                                                Simplex(
                                                    ExactLengthSimplex {
                                                        unit: Px,
                                                        value: NotNan(
                                                            200.0,
                                                        ),
                                                    },
                                                ),
                                            ),
                                        },
                                        Rule {
                                            value: StringValue(
                                                "matrix3d(1,0,0,0,0,1,0,0,0,0,1,0,38,-180,0,1)",
                                            ),
                                        },
                                    ],
                                    updated_at: [
                                        "emg_layout/src/calc.rs:153:29",
                                        "emg_layout/src/calc.rs:153:33",
                                        "emg_layout/src/calc.rs:153:52",
                                        "emg_layout/src/calc.rs:153:71",
                                    ],
                                    pseudo: None,
                                    media: None,
                                    name: "",
                                    keyframes: Keyframes {
                                        frames: [],
                                    },
                                    combinator: None,
                                    pre_combinators: [],
                                },
                            ),
                        },
                    ),
                },
            ),
        },
    ),
    store: RefCell {
        value: the graph store,
    },
}
